File: CSES/Additional Problems/P1655.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class P1655{
   
   public static void main(String[] args)throws IOException{
      FastScanner fs = new FastScanner();
      PrintWriter out = new PrintWriter(System.out);
      
      int n = fs.nextInt();
      int P = 29;
      
      int[] array = fs.nextInts(n);
      
      int[] pxor = new int[n+1];
      pxor[0] = 0;
      for(int k = 0; k < n; k++){
         pxor[k+1] = pxor[k] ^ array[k];
      }
      
      Trie head = new Trie();
      
      for(int k = 0; k <= n; k++){
         Trie cur = head;
         for(int p = P; p >= 0; p--){
            int i = (pxor[k] >> p)&1;
            if(cur.children[i] == null){
               cur.children[i] = new Trie();
            }
            cur = cur.children[i];
         }
      }
      
      int answer = 0;
      for(int k = 1; k <= n; k++){
         Trie cur = head;
         int curanswer = 0;
         for(int p = P; p >= 0; p--){
            int i = (pxor[k] >> p)&1;
            if(cur.children[1-i] != null){
               curanswer += (1 << p);
               cur = cur.children[1-i];
            } else {
               cur = cur.children[i];
            }
         }
         
         answer = Math.max(answer,curanswer);
      }
      
      out.println(answer);
      
      
      
      
      
      
      
      
      
      out.close();
   }
   
   public static class Trie{
      public Trie[] children;
      
      public Trie(){
         children = new Trie[2];
      }
   }
   
      
}

class FastScanner
{
    //I don't understand how this works lmao
    private int BS = 1 << 16;
    private char NC = (char) 0;
    private byte[] buf = new byte[BS];
    private int bId = 0, size = 0;
    private char c = NC;
    private double cnt = 1;
    private BufferedInputStream in;
 
    public FastScanner() {
        in = new BufferedInputStream(System.in, BS);
    }
 
    public FastScanner(String s) {
        try {
            in = new BufferedInputStream(new FileInputStream(new File(s)), BS);
        } catch (Exception e) {
            in = new BufferedInputStream(System.in, BS);
        }
    }
 
    private char getChar() {
        while (bId == size) {
            try {
                size = in.read(buf);
            } catch (Exception e) {
                return NC;
            }
            if (size == -1) return NC;
            bId = 0;
        }
        return (char) buf[bId++];
    }
 
    public int nextInt() {
        return (int) nextLong();
    }
 
    public int[] nextInts(int N) {
        int[] res = new int[N];
        for (int i = 0; i < N; i++) {
            res[i] = (int) nextLong();
        }
        return res;
    }
 
    public long[] nextLongs(int N) {
        long[] res = new long[N+1];
        for (int i = 1; i <= N; i++) {
            res[i] = nextLong();
        }
        return res;
    }
 
    public long nextLong() {
        cnt = 1;
        boolean neg = false;
        if (c == NC) c = getChar();
        for (; (c < '0' || c > '9'); c = getChar()) {
            if (c == '-') neg = true;
        }
        long res = 0;
        for (; c >= '0' && c <= '9'; c = getChar()) {
            res = (res << 3) + (res << 1) + c - '0';
            cnt *= 10;
        }
        return neg ? -res : res;
    }
 
    public double nextDouble() {
        double cur = nextLong();
        return c != '.' ? cur : cur + nextLong() / cnt;
    }
 
    public double[] nextDoubles(int N) {
        double[] res = new double[N];
        for (int i = 0; i < N; i++) {
            res[i] = nextDouble();
        }
        return res;
    }
 
    public String next() {
        StringBuilder res = new StringBuilder();
        while (c <= 32) c = getChar();
        while (c > 32) {
            res.append(c);
            c = getChar();
        }
        return res.toString();
    }
 
    public String nextLine() {
        StringBuilder res = new StringBuilder();
        while (c <= 32) c = getChar();
        while (c != '\n') {
            res.append(c);
            c = getChar();
        }
        return res.toString();
    }
 
    public boolean hasNext() {
        if (c > 32) return true;
        while (true) {
            c = getChar();
            if (c == NC) return false;
            else if (c > 32) return true;
        }
    }
}

File: CSES/Additional Problems/P1655.cpp
#include <bits/stdc++.h>

using namespace std;

struct Trie{
   Trie* children[2];
};

int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   int n;
   cin >> n;
   int P = 30;
   
   vector<int> array(n);
   for(int k = 0; k < n; k++){
      cin >> array[k];
   }
   
   vector<int> pxor(n+1,0);
   for(int k = 0; k < n; k++){
      pxor[k+1] = pxor[k] ^ array[k];
   }
   
   Trie* head = new Trie();
   
   for(int k = 0; k <= n; k++){
      Trie* cur = head;
      for(int p = P; p >= 0; p--){
         int i = (pxor[k] >> p)&1;
         if(cur->children[i] == nullptr){
            cur->children[i] = new Trie();
         }
         cur = cur->children[i];
      }
   }
   
   int answer = 0;
   for(int k = 1; k <= n; k++){
      int curanswer = 0;
      Trie* cur = head;
      for(int p = P; p >= 0; p--){
         int i = (pxor[k] >> p)&1;
         if(cur->children[1-i] != nullptr){
            curanswer ^= (1 << p);
            cur = cur->children[1-i];
         } else {
            cur = cur->children[i];
         }
      }
      
      answer = max(answer,curanswer);
   }
   
   cout << answer << "\n";
   
   
      
   
   
   return 0;
}